File: windows/Max.Core/AgentLoop.cs
using System.Runtime.CompilerServices;
using System.Text.Json.Nodes;

namespace Max.Core;

/// <summary>
/// The agentic loop, ported from the macOS AgentLoop:
/// user message → stream LLM → run tool calls → feed results back → repeat until
/// a turn produces no tool calls. Emits <see cref="AgentEvent"/>s for the UI and
/// appends to the session transcript as it goes.
/// </summary>
public static class AgentLoop
{
    private const int MaxIterations = 30;
    private static readonly HashSet<string> NeedsApproval = new(StringComparer.OrdinalIgnoreCase)
    { "exec", "powershell", "remote_exec", "control_app" };

    public static ILLMProvider CreateProvider(MaxConfig config) => config.Provider switch
    {
        LLMProviderKind.Anthropic => new AnthropicProvider(),
        LLMProviderKind.OpenAI => new OpenAIProvider(),
        LLMProviderKind.Ollama => new OpenAIProvider(NormalizeOllama(config.OllamaBaseURL), requiresKey: false),
        _ => new AnthropicProvider(),
    };

    public static List<IMaxTool> BaseTools(MaxConfig config) =>
        new() { new ExecTool(config), new ReadFileTool(), new WriteFileTool(), new LoopTool() };

    /// <param name="tools">The full tool set (base tools + any platform tools from the head).</param>
    /// <param name="approval">Ask-mode gate. Returns true to allow. Null = auto-approve.</param>
    public static async IAsyncEnumerable<AgentEvent> RunAsync(
        ChatSession session, string userText, MaxConfig config, ISecretStore secrets,
        IReadOnlyList<IMaxTool> tools,
        Func<string, Task<bool>>? approval = null,
        bool isLoopRun = false, bool isRemoteOrigin = false,
        IReadOnlyList<ImagePayload>? images = null,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        if (config.Paused)
        {
            yield return new AgentEvent.Failed("Max is paused. Resume from the tray icon to use it.");
            yield break;
        }
        if (SpendTracker.IsOverLimit(config))
        {
            yield return new AgentEvent.Failed($"Daily spend limit (${config.DailySpendLimitUSD:0.00}) reached. Raise it in Settings.");
            yield break;
        }

        var includeVision = isLoopRun ? (config.AllowScreenVision && config.LoopsCanSeeScreen) : config.AllowScreenVision;
        var registry = new ToolRegistry(tools);
        var provider = CreateProvider(config);
        var apiKey = secrets.ApiKey(config.Provider) ?? "";
        var system = SystemPrompt.Build(config, isLoopRun, includeVision);

        var firstBlocks = new List<ContentBlock> { new ContentBlock.Text(userText) };
        if (images != null) firstBlocks.AddRange(images.Select(i => new ContentBlock.Image(i)));
        session.Append(new ChatTurn(Role.User, firstBlocks));

        var finalText = "";

        for (var iter = 0; iter < MaxIterations; iter++)
        {
            ct.ThrowIfCancellationRequested();

            IReadOnlyList<ContentBlock>? completedBlocks = null;
            var stopReason = "end_turn";

            await using var e = provider
                .StreamTurnAsync(system, PruneImages(session.Turns), registry.Specs, config.Model, apiKey, ct)
                .GetAsyncEnumerator(ct);

            while (true)
            {
                ProviderEvent? ev = null;
                var moved = false;
                var canceled = false;
                string? streamError = null;
                try
                {
                    moved = await e.MoveNextAsync();
                    if (moved) ev = e.Current;
                }
                catch (OperationCanceledException) { canceled = true; }
                catch (Exception ex) { streamError = ex.Message; }

                if (canceled) yield break;
                if (streamError != null) { yield return new AgentEvent.Failed(streamError); yield break; }
                if (!moved || ev is null) break;

                switch (ev)
                {
                    case ProviderEvent.TextDelta td:
                        yield return new AgentEvent.TextDelta(td.Text);
                        break;
                    case ProviderEvent.TurnCompleted tc:
                        completedBlocks = tc.Blocks;
                        stopReason = tc.StopReason;
                        break;
                    case ProviderEvent.Usage u:
                        SpendTracker.Record(config.Model, u.InputTokens, u.OutputTokens);
                        break;
                }
            }

            if (completedBlocks is null)
            {
                yield return new AgentEvent.Failed("model stream ended unexpectedly");
                yield break;
            }

            session.Append(new ChatTurn(Role.Assistant, completedBlocks));
            foreach (var b in completedBlocks)
                if (b is ContentBlock.Text t) finalText = t.Value;

            var toolUses = completedBlocks.OfType<ContentBlock.ToolUse>().ToList();
            if (stopReason != "tool_use" || toolUses.Count == 0) break;

            var results = new List<ContentBlock>();
            foreach (var use in toolUses)
            {
                ct.ThrowIfCancellationRequested();
                var input = ParseObject(use.InputJson);
                var tool = registry.Get(use.Name);
                if (tool is null)
                {
                    results.Add(new ContentBlock.ToolResult(use.Id, $"unknown tool {use.Name}", true, Array.Empty<ImagePayload>()));
                    continue;
                }

                var summary = tool.Summary(input);
                yield return new AgentEvent.ToolStarted(use.Name, summary);

                if (config.ExecApproval == ExecApprovalMode.Ask && !isLoopRun && !isRemoteOrigin
                    && NeedsApproval.Contains(use.Name) && approval != null)
                {
                    var ok = await approval(summary);
                    if (!ok)
                    {
                        ActionLog.Write($"DENIED {use.Name}: {summary}");
                        results.Add(new ContentBlock.ToolResult(use.Id, "User denied this command.", true, Array.Empty<ImagePayload>()));
                        yield return new AgentEvent.ToolFinished(use.Name, "denied", true);
                        continue;
                    }
                }

                ToolOutcome outcome;
                try { outcome = await tool.ExecuteAsync(input, ct); }
                catch (Exception ex) { outcome = ToolOutcome.Fail(ex.Message); }

                ActionLog.Write($"{(outcome.IsError ? "ERROR" : "ran")} {use.Name}: {summary}");
                results.Add(new ContentBlock.ToolResult(use.Id, outcome.Content, outcome.IsError, outcome.Images));
                yield return new AgentEvent.ToolFinished(use.Name,
                    outcome.Content.Length > 200 ? outcome.Content[..200] : outcome.Content, outcome.IsError);
            }

            session.Append(new ChatTurn(Role.User, results));
        }

        yield return new AgentEvent.TurnEnded(finalText);
    }

    /// <summary>Keep only the newest <paramref name="keep"/> image-bearing tool results on the wire.</summary>
    private static IReadOnlyList<ChatTurn> PruneImages(IReadOnlyList<ChatTurn> turns, int keep = 2)
    {
        var imageTurnIndices = new List<int>();
        for (var i = 0; i < turns.Count; i++)
            if (turns[i].Blocks.Any(b => b is ContentBlock.ToolResult r && r.Images.Count > 0))
                imageTurnIndices.Add(i);

        var drop = imageTurnIndices.Take(Math.Max(0, imageTurnIndices.Count - keep)).ToHashSet();
        if (drop.Count == 0) return turns;

        var outList = new List<ChatTurn>(turns.Count);
        for (var i = 0; i < turns.Count; i++)
        {
            if (!drop.Contains(i)) { outList.Add(turns[i]); continue; }
            var blocks = turns[i].Blocks.Select(b => b is ContentBlock.ToolResult r && r.Images.Count > 0
                ? new ContentBlock.ToolResult(r.ToolUseId, r.Content + "\n[screenshot omitted from context — take a fresh one if needed]", r.IsError, Array.Empty<ImagePayload>())
                : b);
            outList.Add(new ChatTurn(turns[i].Role, blocks));
        }
        return outList;
    }

    private static JsonObject ParseObject(string json)
    {
        try { return JsonNode.Parse(string.IsNullOrWhiteSpace(json) ? "{}" : json)?.AsObject() ?? new JsonObject(); }
        catch { return new JsonObject(); }
    }

    private static string NormalizeOllama(string baseUrl)
    {
        var b = baseUrl.Trim().TrimEnd('/');
        return $"{b}/v1/chat/completions";
    }
}

File: Sources/AskMax/Core/AgentLoop.swift
import Foundation
import AppKit

/// The agentic loop, ported from OpenClaw's agent-core design:
/// user message → LLM stream → execute tool calls → feed results back → repeat
/// until a turn produces no tool calls.
enum AgentLoop {

    static func makeRegistry(includeVision: Bool, devices: [RemoteDevice]) -> ToolRegistry {
        var tools: [MaxTool] = [
            ExecTool(),
            AppleScriptTool(),
            ReadFileTool(),
            WriteFileTool(),
            LoopTool(),
        ]
        if includeVision {
            tools.append(SeeScreenTool())
            tools.append(ReadScreenTextTool())
        }
        if !devices.isEmpty {
            tools.append(RemoteExecTool(devices: devices))
        }
        return ToolRegistry(tools)
    }

    /// Old screenshots cost full image tokens on every subsequent request.
    /// Keep only the newest `keep` image-bearing tool results on the wire;
    /// the JSONL transcript on disk keeps everything.
    static func pruneImages(_ turns: [ChatTurn], keep: Int = 2) -> [ChatTurn] {
        var imageTurnIndices: [Int] = []
        for (i, turn) in turns.enumerated() {
            if turn.blocks.contains(where: {
                if case .toolResult(_, _, _, let images) = $0 { return !images.isEmpty }
                return false
            }) {
                imageTurnIndices.append(i)
            }
        }
        let dropIndices = Set(imageTurnIndices.dropLast(keep))
        guard !dropIndices.isEmpty else { return turns }

        return turns.enumerated().map { i, turn in
            guard dropIndices.contains(i) else { return turn }
            var copy = turn
            copy.blocks = turn.blocks.map { block in
                if case .toolResult(let id, let content, let isError, let images) = block, !images.isEmpty {
                    return .toolResult(
                        toolUseId: id,
                        content: content + "\n[screenshot omitted from context — take a fresh one if needed]",
                        isError: isError,
                        images: []
                    )
                }
                return block
            }
            return copy
        }
    }

    static func provider(config: MaxConfig) -> LLMProvider {
        switch config.provider {
        case .anthropic:
            return AnthropicProvider()
        case .openai:
            return OpenAIProvider()
        case .ollama:
            var base = config.ollamaBaseURL.trimmingCharacters(in: .whitespaces)
            while base.hasSuffix("/") { base.removeLast() }
            let url = URL(string: "\(base)/v1/chat/completions")
                ?? URL(string: "http://127.0.0.1:11434/v1/chat/completions")!
            return OpenAIProvider(endpoint: url, requiresKey: false)
        }
    }

    /// Runs one full agentic exchange. Events stream back for the UI;
    /// the session transcript is appended as the loop progresses.
    static func run(
        session: ChatSession,
        userText: String,
        config: MaxConfig,
        isLoopRun: Bool = false,
        isRemoteOrigin: Bool = false
    ) -> AsyncStream<AgentEvent> {
        AsyncStream { continuation in
            let task = Task {
                let includeVision = isLoopRun
                    ? (config.allowScreenVision && config.loopsCanSeeScreen)
                    : config.allowScreenVision
                let registry = makeRegistry(includeVision: includeVision, devices: config.devices)
                let llm = provider(config: config)
                let apiKey = SecretStore.apiKey(for: config.provider) ?? ""
                let system = SystemPrompt.build(
                    config: config, soul: Soul.load(), isLoopRun: isLoopRun, includeVision: includeVision
                )

                // Kill switch: when paused, no tool runs at all.
                if config.paused {
                    continuation.yield(.failed("Max is paused. Resume from the menu-bar icon to use it."))
                    continuation.finish()
                    return
                }
                // Spend cap.
                if SpendTracker.isOverLimit(config) {
                    let limit = String(format: "%.2f", config.dailySpendLimitUSD)
                    continuation.yield(.failed("Daily spend limit ($\(limit)) reached. Raise it in Settings → Spending."))
                    continuation.finish()
                    return
                }

                session.append(ChatTurn(role: .user, blocks: [.text(userText)]))

                var iterations = 0
                var finalText = ""

                loop: while iterations < 30 {
                    iterations += 1
                    var completed: (blocks: [ContentBlock], stopReason: String)?

                    do {
                        let stream = llm.streamTurn(
                            system: system,
                            turns: pruneImages(session.turns),
                            tools: registry.specs,
                            model: config.model,
                            apiKey: apiKey
                        )
                        for try await event in stream {
                            if Task.isCancelled { break loop }
                            switch event {
                            case .textDelta(let piece):
                                continuation.yield(.textDelta(piece))
                            case .turnCompleted(let blocks, let stopReason):
                                completed = (blocks, stopReason)
                            case .usage(let inputTokens, let outputTokens):
                                SpendTracker.record(model: config.model,
                                                    inputTokens: inputTokens, outputTokens: outputTokens)
                            case .toolUseStarted, .toolInputDelta:
                                break // surfaced after input JSON is complete
                            }
                        }
                    } catch {
                        continuation.yield(.failed(error.localizedDescription))
                        continuation.finish()
                        return
                    }

                    guard let turn = completed else {
                        continuation.yield(.failed("model stream ended unexpectedly"))
                        continuation.finish()
                        return
                    }

                    session.append(ChatTurn(role: .assistant, blocks: turn.blocks))
                    for case .text(let t) in turn.blocks { finalText = t }

                    let toolUses = turn.blocks.compactMap { block -> (String, String, String)? in
                        if case .toolUse(let id, let name, let inputJSON) = block {
                            return (id, name, inputJSON)
                        }
                        return nil
                    }

                    if turn.stopReason != "tool_use" || toolUses.isEmpty {
                        break loop
                    }

                    var results: [ContentBlock] = []
                    for (id, name, inputJSON) in toolUses {
                        if Task.isCancelled { break loop }
                        let input = JSON.decode(inputJSON) ?? [:]
                        guard let tool = registry.tool(named: name) else {
                            results.append(.toolResult(toolUseId: id, content: "unknown tool \(name)", isError: true, images: []))
                            continue
                        }
                        let summary = tool.summary(input: input)
                        continuation.yield(.toolStarted(name: name, summary: summary))

                        // Commands that act on the system need approval in ask
                        // mode — but only when a human is at the Mac (the desk
                        // session). Headless loop/channel runs can't show a
                        // dialog; the denylist is their safety net.
                        let needsApproval = ["exec", "applescript", "remote_exec"].contains(name)
                        if config.execApproval == .ask, !isLoopRun, !isRemoteOrigin, needsApproval {
                            let approved = await requestApproval(summary: summary)
                            if !approved {
                                ActionLog.write("DENIED \(name): \(summary)")
                                results.append(.toolResult(toolUseId: id, content: "User denied this command.", isError: true, images: []))
                                continuation.yield(.toolFinished(name: name, resultPreview: "denied", isError: true))
                                continue
                            }
                        }

                        let outcome = await tool.execute(input: input)
                        ActionLog.write("\(outcome.isError ? "ERROR" : "ran") \(name): \(summary)")
                        results.append(.toolResult(
                            toolUseId: id, content: outcome.content,
                            isError: outcome.isError, images: outcome.images
                        ))
                        continuation.yield(.toolFinished(
                            name: name,
                            resultPreview: String(outcome.content.prefix(200)),
                            isError: outcome.isError
                        ))
                    }
                    session.append(ChatTurn(role: .user, blocks: results))
                }

                continuation.yield(.turnEnded(finalText: finalText))
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    @MainActor
    private static func requestApproval(summary: String) async -> Bool {
        let alert = NSAlert()
        alert.messageText = "Max wants to run a command"
        alert.informativeText = summary
        alert.addButton(withTitle: "Allow")
        alert.addButton(withTitle: "Deny")
        alert.alertStyle = .warning
        NSApp.activate(ignoringOtherApps: true)
        return alert.runModal() == .alertFirstButtonReturn
    }
}
